server/src/main/java/org/opensearch/gateway/ShardsBatchGatewayAllocator.java

```diff
@@ -0,0 +1,734 @@
+                shardsToAddToCurrentBatch.clear();
+                batchSize = maxBatchSize;
+            }
+        }
+        return batchesToBeAssigned;
+    }
+
+    private void addBatch(ShardsBatch shardsBatch, boolean primary) {
+        ConcurrentMap<String, ShardsBatch> batches = primary ? batchIdToStartedShardBatch : batchIdToStoreShardBatch;
+        if (batches.containsKey(shardsBatch.getBatchId())) {
+            throw new IllegalStateException("Batch already exists. BatchId = " + shardsBatch.getBatchId());
+        }
+        batches.put(shardsBatch.getBatchId(), shardsBatch);
+    }
+
+    /**
+     * Safely remove a shard from the appropriate batch depending on if it is primary or replica
+     * If the shard is not in a batch, this is a no-op.
+     * Cleans the batch if it is empty after removing the shard.
+     * This method should be called when removing the shard from the batch instead {@link ShardsBatch#removeFromBatch(ShardRouting)}
+     * so that we can clean up the batch if it is empty and release the fetching resources
+     *
+     * @param shardRouting shard to be removed
+     */
+    protected void safelyRemoveShardFromBatch(ShardRouting shardRouting) {
+        String batchId = shardRouting.primary() ? getBatchId(shardRouting, true) : getBatchId(shardRouting, false);
+        if (batchId == null) {
+            logger.debug("Shard[{}] is not batched", shardRouting);
+            return;
+        }
+        ConcurrentMap<String, ShardsBatch> batches = shardRouting.primary() ? batchIdToStartedShardBatch : batchIdToStoreShardBatch;
+        ShardsBatch batch = batches.get(batchId);
+        batch.removeFromBatch(shardRouting);
+        deleteBatchIfEmpty(batches, batchId);
+    }
+
+    /**
+     * Safely remove shard from both the batches irrespective of its primary or replica,
+     * For the corresponding shardId. The method intends to clean up the batch if it is empty
+     * after removing the shard
+     * @param shardRouting shard to remove
+     */
+    protected void safelyRemoveShardFromBothBatch(ShardRouting shardRouting) {
+        String primaryBatchId = getBatchId(shardRouting, true);
+        String replicaBatchId = getBatchId(shardRouting, false);
+        if (primaryBatchId == null && replicaBatchId == null) {
+            return;
+        }
+        if (primaryBatchId != null) {
+            ShardsBatch batch = batchIdToStartedShardBatch.get(primaryBatchId);
+            batch.removeFromBatch(shardRouting);
+            deleteBatchIfEmpty(batchIdToStartedShardBatch, primaryBatchId);
+        }
+        if (replicaBatchId != null) {
+            ShardsBatch batch = batchIdToStoreShardBatch.get(replicaBatchId);
+            batch.removeFromBatch(shardRouting);
+            deleteBatchIfEmpty(batchIdToStoreShardBatch, replicaBatchId);
+        }
+    }
+
+    private void deleteBatchIfEmpty(ConcurrentMap<String, ShardsBatch> batches, String batchId) {
```

Comment:
even this method. Every time a batch is processed, it will get deleted automatically

Reply:
Even before we used to remove entry from ConcurrentHashMap when shard-started event comes, so kept the behaviour intact. Happy to discuss more to change this behaviour.

What we can try is - once cache data is used, whole cache will be cleared for the batch and batch will be removed and reroute will be triggered based on failed nodes or failed shards. 
This can end up creating batches many times. For example if something fails after `cache.getCachedData`, may be in shard allocator or in Allocation Service, today it doesn't require a new fetch, as cache already has the data. In those case, we may end up doing re-fetch. 
For happy case scenarios it should be fine.